Source/Dafny/Resolver/BitvectorOptimization.cs

```diff
@@ -0,0 +1,43 @@
+// Copyright by the contributors to the Dafny Project
+// SPDX-License-Identifier: MIT
+
+using System;
+using System.Linq.Expressions;
+using System.Numerics;
+using Microsoft.Boogie;
+
+namespace Microsoft.Dafny;
+
+public class BitvectorOptimization : IRewriter {
+  public BitvectorOptimization(ErrorReporter reporter) : base(reporter) { }
+
+  internal override void PostResolveIntermediate(ModuleDefinition m) {
+    var visitor = new BitvectorOptimizationVisitor();
+    foreach (var decl in ModuleDefinition.AllCallables(m.TopLevelDecls)) {
+      visitor.Visit(decl);
+    }
+  }
+}
+
+public class BitvectorOptimizationVisitor : BottomUpVisitor {
+  private bool IsShiftOp(BinaryExpr.Opcode op) {
+    return op is BinaryExpr.Opcode.LeftShift or BinaryExpr.Opcode.RightShift;
+  }
+
+  private Expression ShrinkBitVectorShiftAmount(Expression expr, BitvectorType originalType) {
+    var width = new BigInteger(originalType.Width);
+    var intermediateType = new BitvectorType((int)width.GetBitLength());
+    var newExpr = new ConversionExpr(expr.tok, expr, intermediateType, "when converting shift amount to a bit vector, the ");
```

Comment:
This error message doesn't occur in the new test cases, and I didn't find it in existing ones. Shouldn't we test for this message?